File: mm.py
"""
Caroline Danzi
Dr. Zmuda
CSE 465

Note: I used http://www.tutorialspoint.com/python/python_command_line_arguments.htm
to learn how to get command line arguments in Python.  I also used the Python 3
documentation for string methods, and notes from another class for regex syntax.
"""

import sys
import re

# The first command line argument at 0 is the file name, the second will be
# the tsv, and the third will be the tmp file.  
tsvFileName = sys.argv[1]
tmpFileName = sys.argv[2]

# open the files for reading
tsvFile = open(tsvFileName, 'r')

# The first line in the tsv file has the column headings
colHeadings = tsvFile.readline().strip().split('\t')

# Build a dictionary of the column headings where the key
# is the string heading and the value is the index
columns = {}
i = 0
for column in colHeadings:
    columns[column] = i
    i += 1

# Get the index of the ID column for file name
idIndex = columns['ID']

for row in tsvFile:
    cells = row.strip().split('\t')
    fileName = cells[idIndex] + '.txt'
    file = open(fileName, 'w')
    tmpFile = open(tmpFileName, 'r')

    # Go through each line in the template file, see if it has
    # any <<tags>>, replace them, and write the string to the output file
    for line in tmpFile:
        pattern = re.compile('<<(.*?)>>')
        # finditer returns a sequence of match objects
        matchIterator = pattern.finditer(line)
        for match in matchIterator:
            header = match.group(1)
            attribute = cells[columns[header]]
            line = re.sub('<<' + header + '>>', attribute, line)
        file.write(line)
    file.close()

tsvFile.close()
tmpFile.close()

"""
.search to .finditer or .findall
"""
    
        
    

File: mm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.IO;

namespace CSE465 {

	public class MailMerge {
		
		private string tsvFile, tmpFile;
		private Dictionary<string, int> headers;
		
		public MailMerge(string tsvFile, string tmpFile) {
			this.tsvFile = tsvFile;
			this.tmpFile = tmpFile;
			this.headers = new Dictionary<string, int>();
		}
		
		public void parseTsv() {
			StreamReader tsvReader = new StreamReader(tsvFile);
			
			// The first line has the headers, so create
			// the dictionary of header and index pairs
			string headerLine = tsvReader.ReadLine();
			string[] tsvHeaders = headerLine.Split('\t');
			for(int i = 0; i < tsvHeaders.Length; i++) {
				headers.Add(tsvHeaders[i], i);
			}
			
			int idIndex = headers["ID"];
			StreamWriter output;
			StreamReader tmpReader;
			// Go through each line in the tsvFile
			while(!tsvReader.EndOfStream) {
				// Get the line, split it based on tab
				string line = tsvReader.ReadLine();
				string[] tokens = line.Split('\t');
				
				// Create and open the file for writing
				string fileName = tokens[idIndex] + ".txt";
				output = new StreamWriter(fileName);
				tmpReader = new StreamReader(tmpFile);
				
				while(!tmpReader.EndOfStream) {
					string templateLine = tmpReader.ReadLine();
					line = replaceHeaders(templateLine, tokens);
					output.WriteLine(line);
				}
						
				output.Close();
				tmpReader.Close();
			}
			
		}
		
		private string replaceHeaders(string line, string[] tokens) {
			string pattern = "<<(.*?)>>";
			Regex rx = new Regex(pattern, RegexOptions.IgnoreCase);
			MatchCollection matches = rx.Matches(line);
			
			foreach(Match m in matches) {
				string attribute = tokens[headers[m.Groups[1].ToString()]];
				line = line.Replace(m.Value, attribute);
			}
			return line;
		}
		
		public static void Main(string[] args) {
			MailMerge mm = new MailMerge(args[0], args[1]);
			mm.parseTsv();
		}
	}

}